components/script/dom/node.rs

```diff
@@ -1283,37 +1258,22 @@ impl Iterator for PrecedingNodeIterator {
     }
 }
 
-pub struct LastChildIterator {
-    current: Option<Root<Node>>,
-}
-
-impl Iterator for LastChildIterator {
-    type Item = Root<Node>;
-
-    fn next(&mut self) -> Option<Root<Node>> {
-        let current = match self.current.take() {
-            None => return None,
-            Some(current) => current,
-        };
-        self.current = current.GetLastChild();
-        Some(current)
-    }
-}
-
-pub struct AncestorIterator {
+struct SimpleNodeIterator<I>
+    where I: Fn(&Node) -> Option<Root<Node>>
+{
     current: Option<Root<Node>>,
+    next_node: I,
 }
 
-impl Iterator for AncestorIterator {
+impl<I> Iterator for SimpleNodeIterator<I>
+    where I: Fn(&Node) -> Option<Root<Node>>
+{
     type Item = Root<Node>;
 
-    fn next(&mut self) -> Option<Root<Node>> {
-        let current = match self.current.take() {
-            None => return None,
-            Some(current) => current,
-        };
-        self.current = current.GetParentNode();
-        Some(current)
+    fn next(&mut self) -> Option<Self::Item> {
+        let current = self.current.take();
+        self.current = current.as_ref().and_then(|c| (&self.next_node)(c));
```

Comment:
What's the magic behind `(&self.next_node)`? Specifically, I don't understand why you need the & and the parentheses here.